File: app/src/main/java/com/permissionnanny/BaseReceiver.java
package com.permissionnanny;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.VisibleForTesting;
import com.permissionnanny.dagger.ContextComponent;
import com.permissionnanny.dagger.ContextModule;
import com.permissionnanny.dagger.DaggerContextComponent;

/**
 * The root of all BroadcastReceivers.
 */
public class BaseReceiver extends BroadcastReceiver {

    private ContextComponent mComponent;

    @Override
    public void onReceive(Context context, Intent intent) {}

    @VisibleForTesting
    public void setTestComponent(ContextComponent component) {
        mComponent = component;
    }

    public ContextComponent getComponent(Context context) {
        if (mComponent == null) {
            mComponent = DaggerContextComponent.builder()
                    .appComponent(((App) context.getApplicationContext()).getAppComponent())
                    .contextModule(new ContextModule(context))
                    .build();
        }
        return mComponent;
    }
}

File: app/src/main/kotlin/com/permissionnanny/BaseReceiver.kt
package com.permissionnanny

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.support.annotation.VisibleForTesting
import com.permissionnanny.dagger.ContextComponent
import com.permissionnanny.dagger.ContextModule
import com.permissionnanny.dagger.DaggerContextComponent

/**
 * The root of all BroadcastReceivers.
 */
open class BaseReceiver : BroadcastReceiver() {

    private var component: ContextComponent? = null

    override fun onReceive(context: Context, intent: Intent) {
    }

    @VisibleForTesting
    fun setTestComponent(component: ContextComponent) {
        this.component = component
    }

    fun getComponent(context: Context): ContextComponent {
        if (component == null) {
            component = DaggerContextComponent.builder()
                    .appComponent((context.applicationContext as App).getAppComponent())
                    .contextModule(ContextModule(context))
                    .build()
        }
        return component!!
    }
}
